markets/storageadapter/dealpublisher.go

```diff
@@ -443,3 +467,83 @@ func (p *DealPublisher) filterCancelledDeals() {
 	}
 	p.pending = filtered
 }
+
+func (p *DealPublisher) queueDeals(pdeal *pendingDeal) {
+	p.lk.Lock()
+	defer p.lk.Unlock()
+
+	// Filter out any cancelled deals
+	p.filterCancelledDeals()
+
+	// Make sure the new deal hasn't been cancelled
+	if pdeal.ctx.Err() != nil {
+		return
+	}
+
+	// Add the new deal to the queue
+	p.pending = append(p.pending, pdeal)
+	log.Infof("add deal with piece CID %s to publish deals queue - %d deals in queue",
+		pdeal.deal.Proposal.PieceCID, len(p.pending))
+}
+
+func (p *DealPublisher) PublishQueuedDeals(deals []cid.Cid) (error, []cid.Cid) {
+	p.lk.Lock()
+	var notFound []cid.Cid
+	var toPublish []*pendingDeal
+
+	// Check that each deal is part of the queue
+	for _, c := range deals {
+		found := false
+		for _, pd := range p.pending {
+			if c.Equals(pd.proposalCid) {
+				found = true
+				toPublish = append(toPublish, pd)
+				break
+			}
+		}
+
+		if !found {
+			notFound = append(notFound, c)
+		}
+	}
+
+	p.lk.Unlock()
+
+	if len(notFound) > 0 {
+		return fmt.Errorf("deals not found in the publisher queue"), notFound
+	}
```

Comment:
I don't think we need to return an error here.
I'd suggest that the resolver should return a list of the deals that were published. For example if
- the user wants to publish deals [1, 2, 3]
- deal 2 was not found (maybe it was already published)
- the resolver should return [1, 3]